lib/storage/batch.go

```diff
@@ -0,0 +1,126 @@
+package storage
+
+import (
+	"sync"
+
+	"github.com/syndtr/goleveldb/leveldb"
+	leveldbIterator "github.com/syndtr/goleveldb/leveldb/iterator"
+	leveldbOpt "github.com/syndtr/goleveldb/leveldb/opt"
+	leveldbUtil "github.com/syndtr/goleveldb/leveldb/util"
+)
+
+type BatchCore struct {
+	sync.RWMutex
+
+	core  LevelDBCore
+	batch *leveldb.Batch
+
+	inserted map[string][]byte
+}
+
+func NewBatchCore(core LevelDBCore) *BatchCore {
+	return &BatchCore{
+		core:     core,
+		batch:    &leveldb.Batch{},
+		inserted: map[string][]byte{},
+	}
+}
+
+func (bb *BatchCore) Has(key []byte, opt *leveldbOpt.ReadOptions) (bool, error) {
+	bb.RLock()
+	defer bb.RUnlock()
+
+	var found bool
+	if _, found = bb.inserted[string(key)]; found {
+		return true, nil
+	}
+
+	return bb.core.Has(key, opt)
+}
+
+func (bb *BatchCore) Get(key []byte, opt *leveldbOpt.ReadOptions) (b []byte, err error) {
+	bb.RLock()
+	defer bb.RUnlock()
+
+	var found bool
+	if b, found = bb.inserted[string(key)]; found {
+		return
+	}
+
+	return bb.core.Get(key, opt)
+}
+
+// NewIterator does not work with `BatchBackend`
+func (bb *BatchCore) NewIterator(r *leveldbUtil.Range, opt *leveldbOpt.ReadOptions) leveldbIterator.Iterator {
+	return bb.core.NewIterator(r, opt)
+}
+
+func (bb *BatchCore) Put(key []byte, v []byte, opt *leveldbOpt.WriteOptions) error {
+	bb.Lock()
+	defer bb.Unlock()
+
+	bb.inserted[string(key)] = v
+	bb.batch.Put(key, v)
+
+	return nil
+}
+
+// Write will write the existing contents of `BatchBackend.batch` and then
+// argument, batch will be written.
+func (bb *BatchCore) Write(batch *leveldb.Batch, opt *leveldbOpt.WriteOptions) (err error) {
+	bb.Lock()
+	defer bb.Unlock()
+
+	if err = bb.core.Write(bb.batch, opt); err != nil {
```

Comment:
> @kfangw : I also confused. When the batch argument is needed?

Yes, check https://godoc.org/github.com/syndtr/goleveldb/leveldb#DB.Write .
```
func (db *DB) Write(batch *Batch, wo *opt.WriteOptions) error
```

Reply:
> @kfangw : 
> - Remove News, Sets method from `LevelDBBackend`

Yes, it could be, but I don't want to break the existing code :) We can tidy up in another issue.

Reply:
@spikeekips 

That is `DB`'s Write method.

If the role of the `BatchCore` is clear then it is fine.

Is the `BatchCore` DB? , Batch? or both?

Reply:
The shape of `BatchCore` is almost same with `leveldb.Transaction` and `leveldb.DB`, it is one of `LevelDBCore`.

Reply:
Then, it is `DB` but wraps `Batch` also.
I understand your intention not to effects other code